lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -2926,4 +2926,96 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "failed to squeeze scores tensor");
+        squeezedScores = Torch::squeezeTensor(
+            rewriter, binder.op, binder.getLoc(), 0, squeezedScores.value());
+        if (failed(squeezedBoxes))
+          return rewriter.notifyMatchFailure(binder.op,
+                                             "failed to squeeze scores tensor");
+
+        boxes = squeezedBoxes.value();
+        scores = squeezedScores.value();
+
+        // TODO: Add support for handling max_output_boxes_per_class
+        // and score_threshold arg.
+        // If num_boxes (N) > max_output_boxes_per_class then the op can't be
+        // lowered since the torchvision::nms op doesn't have support for
+        // handling the max_output_boxes_per_class arg.
+        Value maxOutputBoxesPerClass = rewriter.create<Torch::AtenItemOp>(
+            binder.getLoc(), rewriter.getType<Torch::IntType>(), operands[2]);
+        Value cstZero = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getI64IntegerAttr(0));
+        Value numBoxes = rewriter.create<Torch::AtenSizeIntOp>(binder.getLoc(),
+                                                               boxes, cstZero);
+        Value boxesCond = rewriter.create<Torch::AtenGtIntOp>(
+            binder.getLoc(), numBoxes, maxOutputBoxesPerClass);
+        rewriter.create<Torch::RuntimeAssertOp>(
+            binder.getLoc(), boxesCond,
+            rewriter.getStringAttr("unimplemented: number of boxes should be "
+                                   "<= max_output_boxes_per_class"));
+
+        // If score_threshold < max(scores) then the op can't be lowered since
+        // the torchvision::nms op doesn't have support for handling the
+        // score_threshold arg.
+        Value scoreThreshold = rewriter.create<Torch::AtenItemOp>(
+            binder.getLoc(), rewriter.getType<Torch::FloatType>(), operands[4]);
+        Value maxScores = rewriter.create<Torch::AtenMaxOp>(
+            binder.getLoc(),
+            Torch::ValueTensorType::get(binder.op->getContext(), {},
+                                        rewriter.getF32Type()),
+            scores);
+        maxScores = rewriter.create<Torch::AtenItemOp>(
+            binder.getLoc(), rewriter.getType<Torch::FloatType>(), maxScores);
+
+        Value scoresCond = rewriter.create<Torch::AtenGtFloatOp>(
+            binder.getLoc(), maxScores, scoreThreshold);
+        rewriter.create<Torch::RuntimeAssertOp>(
+            binder.getLoc(), scoresCond,
+            rewriter.getStringAttr(
+                "unimplemented: score_threshold should be >= max(scores)"));
+
+        Value iouThreshold = rewriter.create<Torch::AtenItemOp>(
+            binder.getLoc(), rewriter.getType<Torch::FloatType>(), operands[3]);
+        rewriter.replaceOpWithNewOp<Torch::TorchvisionNmsOp>(
+            binder.op, resultType, boxes, scores, iouThreshold);
```

Comment:
The result type of onnx.NonMaxSuppression doesn't match TorchvisionNmsOp. It looks like nms has output shape `[n]` where `n` is the number of selected indices out of `N` (which I assume corresponds to onnx's `spatial_dimension`). The onnx op has output shape `[n, 3]` (the three is the tuple `batch_index`, `channel_index` , and `box_index` -- or I assume `spatial_index` by their docs), so we might need to concatenate the torchvision result with a `torch.zeroes([n,2])`, since this conversion will only work when the batches and channels are both 1.

I'm wondering if these two ops are sufficiently different to consider something else entirely. I know there is a [batched_nms](https://pytorch.org/vision/main/generated/torchvision.ops.batched_nms.html), but it's in a completely different format. For onnx boxes with shape `[B,N]`, we would need to flatten it to shape `[B*N]`, and make a torchvision `idxs` tensor which is like `N` copies of `arange(0,B)` flattened to shape `[B*N]`. Then, in the onnx `scores` tensor of shape `[B, C, N]`, extract each channel one at a time and perform batched_nms. Here is some pseudocode:

```python
flattened_boxes = boxes.flatten()
idxs = arange(0, B).unsqueeze(1).repeat(1, N).flatten()
# or maybe easier:
idxs = floor(arange(0, B, 1/N))
onnx_result = tensor.empty(resultShape)
count = 0
for c in range(C):
   channelscores = scores[ : , c , : ]
   flattenedscores = channelscores.flatten() # shape = [B*N]
   result = torchvision.batched_nms(flattenedboxes, flattenedscores, idxs, iou_threshold)
   #result contains the values indexing into [B*N] corresponding to the selected boxes
   for res_index in result:
      batch_idx = idxs[res_index]
      channel_ix = c
      box_idx = res_index % batch_idx
      onnx_result[count, 0] = batch_idx
      onnx_result[count, 1] = channel_idx
      onnx_result[count, 2] = box_idx
      count++
return onnx_result
```

Frankly, I'm just guessing how these ops line up since the documentation is a bit limited. It might be good to compare something like the python pseudocode implementation above with the onnx runtime result and see if this is even a viable approach.

Reply:
I'd generate an onnx model with the helpers and build an onnxruntime inference session (like we do in the shark-testsuite model.py files).

If you have the python impl to compare with, see if it gives the same results.

Reply:
Hi @zjgarvey, I have modified the original lowering since it's already implicitly checking for B,C to be 1. Hence, we can support this op now for limited cases and can extend it in future, if required.